merkletree.go

```diff
@@ -0,0 +1,344 @@
+package merkletree
+
+import (
+	"crypto"
+	"errors"
+	"hash"
+)
+
+var (
+	ErrInvalidTree = errors.New("[merkletree] invalid tree")
+)
+
+const (
+	EmptyBranchIdentifier = 'E'
+	LeafIdentifier        = 'L'
+)
+
+type Scheme interface {
+	Sign([]byte) []byte
+	Verify(publicKey []byte, message, sig []byte) bool
+}
+
+type HashFunction struct {
+	HashSizeByte int
+	Hash         hash.Hash
+	HashId       crypto.Hash
+}
+
+type MerkleTree struct {
+	hash      HashFunction
+	scheme    Scheme
+	treeNonce []byte
+	salt      []byte
+	root      *interiorNode
+}
+
+type interiorNode struct {
+	parent     *interiorNode
+	leftChild  interface{}
+	rightChild interface{}
+	leftHash   []byte
+	rightHash  []byte
+	level      int
+}
+
+type userLeafNode struct {
+	interiorNode
+	key        string
+	value      []byte
+	index      []byte
+	commitment []byte
+}
+
```

Comment:
I think it may make sense to put all node-related definitions and functions in their own separate node module.